cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -91,6 +89,101 @@ Result<DeclarationInfo> ProcessEmit(const RelMessage& rel,
   }
 }
 
+template <typename RelMessage>
+Result<DeclarationInfo> ProcessEmit(const RelMessage& rel,
+                                    const DeclarationInfo& no_emit_declr,
+                                    const std::shared_ptr<Schema>& schema) {
+  return ProcessEmit(rel.has_common() ? std::optional(rel.common()) : std::nullopt,
+                     no_emit_declr, schema);
+}
+
+Result<DeclarationInfo> ProcessExtensionEmit(
+    const DeclarationInfo& no_emit_declr, const std::vector<int>& emit_info,
+    const std::vector<int>& field_output_indices) {
+  const std::shared_ptr<Schema>& input_schema = no_emit_declr.output_schema;
+  std::vector<compute::Expression> proj_field_refs;
+  proj_field_refs.reserve(emit_info.size());
+  FieldVector emit_fields;
+  emit_fields.reserve(emit_info.size());
+
+  for (int emit_idx : emit_info) {
+    if (emit_idx < 0 || static_cast<size_t>(emit_idx) >= field_output_indices.size()) {
+      return Status::Invalid("Out of bounds emit index ", emit_idx);
+    }
+    int field_idx = field_output_indices[emit_idx];
+    if (field_idx < 0) {
+      return Status::Invalid("Non-output emit index ", emit_idx);
+    }
+    proj_field_refs.push_back(compute::field_ref(FieldRef(field_idx)));
+    emit_fields.push_back(input_schema->field(field_idx));
+  }
+
+  std::shared_ptr<Schema> emit_schema = schema(std::move(emit_fields));
+
+  return DeclarationInfo{
+      compute::Declaration::Sequence(
+          {no_emit_declr.declaration,
+           {"project", compute::ProjectNodeOptions{std::move(proj_field_refs)}}}),
+      std::move(emit_schema)};
+}
+
+Result<RelationInfo> GetExtensionRelationInfo(const substrait::Rel& rel,
+                                              const ExtensionSet& ext_set,
+                                              const ConversionOptions& conv_opts,
+                                              std::vector<DeclarationInfo>& inputs) {
```

Comment:
Our style guide prefers pointers for out parameters instead of mutable references (more obvious at the call site that something is going to be modified)

Reply:
Will fix.